### ISSUE TITLE
Istio routing request using wrong service?

### ISSUE THREAD

**Issue by brackendawson** [istio/istio#21856]
**Bug description**
I believe that when making a request to a ClusterIP service; Istio is applying the policy and destination rule from another service that has the same endpoints.

The setup is:
***monitoring namespace:***
* Sidecar injection disabled.
* A ClusterIP service with a cluster IP called `prometheus-registry` with a Prometheus pod as its only endpoint.
* A ClusterIP service with **no cluster IP** called `prometheus-operated` with the same Prometheus pod as its only endpoint, this service is an artefact of using Prometheus Operator.
* Specific policy for `prometheus-registry.monitoring.svc.cluster.local` to disable mtls
* The Prometheus pod is in this namespace.

***default namespace:***
* Sidecar injection enabled.
* Wide-reaching policy and destination rule for STRICT ISTIO_MUTUAL on `*.local`
* Specific destination rule for `prometheus-registry.monitoring.svc.cluster.local:9090` for DISABLE
* A client pod in this namespace makes requests to `http://prometheus-registry.monitoring.svc.cluster.local:9090`

We noticed that despite our policy and destination rule allowing the client to talk to prometheus-registry on port 9090 without using mtls, connections are immediately reset. Checking the envoy logs from the client pod, I can see that the request to `http://prometheus-registry.monitoring.svc.cluster.local:9090` is instead seen as an outgoing connection to `prometheus-operated.monitoring.svc.cluster.local` on 9090:
> [2020-03-05T16:21:37.200Z] "- - -" 0 UF,URX "-" "-" 0 0 6 - "-" "-" "-" "-" "172.21.232.131:9090" outbound|9090||prometheus-operated.monitoring.svc.cluster.local - 172.21.232.131:9090 172.30.201.101:44790 - -

When we apply the STRICT DISABLE policy and destination rule to the `prometheus-operated` service, we are then able to make requests to the `prometheus-registry` service.

**Expected behavior**
Requests to a named service should be routed to the endpoint/s of that service using the policy and destination rule applying to that service.

**Steps to reproduce the bug**
1. Deploy Prometheus into a monitoring namespace using Prometheus Operator and include a service to access Prometheus.
2. Deploy a client accessing Prometheus into the default namespace.
3. Deploy istio into the cluster.
4. Enable sidecar injection in the default namespace.
5. Set STRICT ISTIO_MUTUAL for all services.
6. Set STRICT DISABLE for your Prometheus service.
7. Make a request to your Prometheus service from a client in the default namespace, hopefully you will see that this request fails and that envoy sees it as a connection to the `prometheus-operated` service rather than your own service.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
```
$ istioctl version --remote
client version: 1.4.5
control plane version: 1.4.5
data plane version: 1.4.5 (138 proxies)
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.2", GitCommit:"59603c6e503c87169aea6106f57b9f242f64df89", GitTreeState:"clean", BuildDate:"2020-01-23T14:21:54Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.7+IKS", GitCommit:"8e6ceab7b61240d0beca0e2a2d78a91f52c5df18", GitTreeState:"clean", BuildDate:"2020-02-13T08:09:12Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**
Managed Istio addon on IBM Cloud Kubernetes Service

**Environment where bug was observed (cloud vendor, OS, etc)**
IBM Cloud Kubernetes Service

**Comment by howardjohn** [comment]
I tried to reproduce on latest code and could not. This area has had substantial improvements, I strongly suspect this has been fixed but cannot point to a specific PR. Can you try on 1.6+? I tried on latest code, its possible its not fixed in 1.6 but is on latest if you want to try that as well, see https://github.com/istio/istio/wiki/Dev%20Builds.
My config:

```yaml
apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  name: mutual
spec:
  host: "*.cluster.com"
  trafficPolicy:
    tls:
      mode: ISTIO_MUTUAL
---
apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  name: mutual
spec:
  host: "httpbin.httpbin"
  trafficPolicy:
    tls:
      mode: DISABLE
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: httpbin
spec:
  selector:
    matchLabels:
      app: httpbin
  template:
    metadata:
      labels:
        app: httpbin
    spec:
      containers:
      - name: httpbin
        image: docker.io/kennethreitz/httpbin
        resources:
          limits:
            memory: "128Mi"
            cpu: "500m"
        ports:
        - containerPort: 80
        readinessProbe:
          failureThreshold: 30
          httpGet:
            path: /get
            port: 80
            scheme: HTTP
          initialDelaySeconds: 4
          periodSeconds: 5
          successThreshold: 1
          timeoutSeconds: 1
---
apiVersion: v1
kind: Service
metadata:
  name: httpbin
spec:
  selector:
    app: httpbin
  ports:
  - name: tcp
    port: 80
    targetPort: 80
---
apiVersion: v1
kind: Service
metadata:
  name: httpbin-headless
spec:
  clusterIP: None
  selector:
    app: httpbin
  ports:
  - name: tcp
    port: 80
    targetPort: 80

```

**Comment by howardjohn** [comment]
I was unable to reproduce and there isn't enough information here to move forward. If you still see issues feel free to reopen. Thanks!
